src/main.rs

```diff
@@ -1,38 +1,80 @@
 use std::env;
 use std::collections::HashSet;
 
-use serenity::{
-    async_trait,
-    http::Http,
-    framework::{standard::macros::group, StandardFramework},
-    model::gateway::Ready,
-    prelude::*
-};
+use serenity::{async_trait, client::bridge::gateway::GatewayIntents, framework::{standard::macros::{group, hook}, StandardFramework}, http::Http, model::channel::Message, model::gateway::Ready, prelude::*, utils::MessageBuilder};
 
-use log::{error, warn, info};
+use chrono::prelude::*;
+
+use tracing::{info, error, warn, instrument};
 
 mod commands;
 
+
+pub mod util;
+pub mod errors;
+
+pub struct RedisConnection;
+impl TypeMapKey for RedisConnection {
+    type Value = redis::Connection;
+}
+
 use commands::meta::*;
+use commands::leveling::*;
 #[group]
 #[commands(ping)]
 struct Meta;
 
+#[group]
+#[commands(rank, levels)]
+struct Leveling;
+
 struct Handler;
 
 #[async_trait]
 impl EventHandler for Handler {
-    async fn ready(&self, _: Context, ready: Ready) {
-        info!("Connected as {}", ready.user.name)
+    async fn ready(&self, _ctx: Context, ready: Ready) {
+        info!("Logged into Discord as {}", ready.user.name)
+    }
+
+    #[instrument(skip(self, ctx))]
+    async fn message(&self, ctx: Context, msg: Message) {
+        if !msg.content.starts_with(&env::var("DISCORD_PREFIX").unwrap()) {
+            if !msg.author.bot {
+                let mut bot_data = ctx.data.write().await;
+                let mut redis_conn = bot_data.get_mut::<RedisConnection>().unwrap();
+                match util::leveling::get_user_level(msg.author.id.0, &mut redis_conn) {
+                    Ok(data) => {
+                        let time_since_last_msg = data.last_msg - Utc::now();
+                        if time_since_last_msg.num_minutes() < -1 {
```

Comment:
Any particular reason for having the time be negative?

Reply:
Durations between the current `DateTime<Utc>` and the previous `DateTime` represented by the last time the member sent a message are negative, because the stored `DateTime` is in the past.